docs/2018-11-12-Unification.md

```diff
@@ -0,0 +1,360 @@
+
+Problem
+-------
+
+* Finding if there are `t(X)`, `S(X, Y)` and `P(X)` such that
+  `φ(X) ∧ α(Y) = t(X) ∧ S(X, Y) ∧ P(X)`, i.e. finding if `φ(X)` and `α(Y)`
+  *unify*.
+* Actually finding those patterns.
+
+Additional constraints and assumptions
+--------------------------------------
+
+* The unification algorithm assumes that it mostly unifies terms.
+* We are only interested in substitutions that resolve the entire `Y`.
+* We will assume that infinite constructor-based recursion can't exist,
+  e.g. `[x = C(x)]` is bottom.
+* We will give up if we have infinite recursion based on things other than
+  constructors, e.g. `[x = f(x)]`.
+* Whenever we give up (i.e. because we don't know how how to handle something),
+  we will return bottom instead of signaling an error.
+
+Solution
+--------
+
+Note that it is not always possible to determine if two patterns unify.
+The algorithm below describes some of the cases when this is possible.
+
+The algorithm works in a top-down fashion, but since unification
+can't always be resolved efficiently locally
+(think `concat(X, Y->Z)` unified with a concrete map, where we'd like
+to resolve the `Y` variable before the unification), we represent
+things that should normally be substitutions as predicates, and, at various
+points in the algorithm, we try to transform them into substitutions.
+The exact points where this happens are left open (current implementation
+does this whenever we merge substitutions and predicates).
+
+The algorithm has a list of algorithms for the cases where it knows how
+to solve unification. Each such algorithm is expected to return:
+* *not-handled-error* if it does not know how to handle the provided patterns
+* *handled-with-error* if it knows how to handle the provided patterns,
+  but there was an error not specific to the current algorithm (e.g. children's
+  unification failed with error).
+* *bottom* if it knows how to handle the patterns, but unification is
+  impossible (e.g. it tries to unify different constructors).
+* a *pattern* of the form `t(X) ∧ S(X, Y) ∧ P(X)` if unification succeeded.
+
+Then the algorithm looks like this:
+
+```
+unify(p1, p2):
+  for unifier in known-unification-algorithms
+    case unifier(p1, p2) of
+      not-handled-error -> continue
+      handled-with-error -> return handled-with-error
+      bottom -> return bottom
+      pattern -> return pattern
+  return handled-with-error  -- counterintuitive, but makes things easier.
+```
+
+Known unification cases
+-----------------------
+
+The order of these cases in the `known-unification-algorithms` list is
+important both for optimization purposes and because, in practice,
+later cases rely on being called only when some of the previous cases fail
+(they could retest the previous cases' conditions, but, for simplicity,
+they don't).
+
+### bool and
+
+Resolves unification with top and bottom patterns.
+
+```
+T ∧ φ = φ
+φ ∧ T = φ
+⊥ ∧ φ = ⊥
+φ ∧ ⊥ = ⊥
+not-handled-error otherwise
+```
```

Comment:
Seems to me that we're sort of conflating between unification and simplification here...

Reply:
I defined unification as "and" that is evaluated to a specific form, and that includes these rules.